pkg/resources/kafka/kafka.go

```diff
@@ -306,7 +301,36 @@ func (r *Reconciler) Reconcile(log logr.Logger) error {
 		}
 	}
 
+	// all broker nodes under the same Kafka cluster must use the same cluster UUID
+	if r.KafkaCluster.Spec.KRaftMode && r.KafkaCluster.Status.ClusterID == "" {
+		r.KafkaCluster.Status.ClusterID = generateRandomClusterID()
+		err = r.Client.Status().Update(ctx, r.KafkaCluster)
+		if apierrors.IsNotFound(err) {
+			err = r.Client.Update(ctx, r.KafkaCluster)
+		}
+		if err != nil {
+			return errors.NewWithDetails("could not update ClusterID status",
+				"component", componentName,
+				"clusterName", r.KafkaCluster.Name,
+				"clusterNamespace", r.KafkaCluster.Namespace)
+		}
+	}
+
+	controllerID, err := r.determineControllerId()
+	if err != nil {
+		log.Error(err, "could not find controller broker")
+	}
+
+	// In KRaft mode:
+	// 1. there is no way for admin client to know which node is the active controller, controllerID obtained above is just a broker ID of a random active broker (this is intentional by Kafka)
+	// 2. the follower controllers replicate the data that is written to the active controller and serves as hot standbys if the active controller fails.
+	//    Because the controllers now all track the latest state, controller fail-over will not require a lengthy reloading time to have all the state to transfer to the new controller
+	// Therefore, by setting the controllerID to be -1 to not take the controller identity into consideration when reordering the brokers
+	if r.KafkaCluster.Spec.KRaftMode {
+		controllerID = -1
+	}
```

Comment:
The potential optimization that we can do here is to ensure there at least one controller broker gets up first during the cluster start-up process, this way the normal brokers can register themselves to the controller broker as soon as they are ready to send out the registration requests.

Currently, if all the normal brokers are reordered in the front of the reconciliation queue and we are in a huge Kafka cluster (let's say 997 brokers + 3 controllers), the 997 normal brokers would need to wait for >=1 of the three controllers to be ready so they can register themselves; and if the K8s cluster doesn't have enough resources to the 998th broker pod in the list (which is the first controller node), the entire Kafka cluster will in a completely non-functional state.

Since this doesn't really affect the main logic, I'd like to do this part in a separate PR so this PR doesn't keep growing